### PR TITLE
do_patch & do_wget: Rework to always check if local=remote

### LINKED ISSUE #1369: Building ffmpeg not working - libsvt_hevc error
```
10:45:56   Running git update for ffmpeg...
10:45:58 ┌ ffmpeg git  ............................... [Recently updated]
10:45:58 ├ Until SVT-AV1 issues a fix, libsvtav1 must be disabled while libsvthevc is enabled....
10:45:58 ├ Changing options to comply to nonfree...
CUDA_PATH environment variable not set or directory does not exist.
10:45:59 Installing clang... done
10:46:04 ├ Compiling static FFmpeg...
10:46:06 ├ Running configure...
10:51:09 ├ Running make...
Likely error:
CC      libavfilter/vf_scale_cuda.ptx.o
CC      libavfilter/vf_thumbnail_cuda.ptx.o
CC      libavfilter/vf_yadif_cuda.ptx.o
AR      libavfilter/libavfilter.a
CC      libavcodec/libsvt_hevc.o
C:/Development/MABS/build/ffmpeg-git/libavcodec/libsvt_hevc.c:24:10: fatal error: EbTime.h: No such file or directory
   24 | #include "EbTime.h"
      |          ^~~~~~~~~~
compilation terminated.
make: *** [/build/ffmpeg-git/ffbuild/common.mak:59: libavcodec/libsvt_hevc.o] Error 1
make failed. Check C:/Development/MABS/build/ffmpeg-git/build-static-64bit/ab-suite.make.log
This is required for other packages, so this script will exit.
11:05:26   Creating diagnostics file...

All relevant logs have been anonymously uploaded to https://0x0.st/z4NC.zip
Copy and paste [logs.zip](https://0x0.st/z4NC.zip) in the GitHub issue.
Make sure the suite is up-to-date before reporting an issue. It might've been fixed already.
Try running the build again at a later time.

```

### PR DESCRIPTION
**Pull request by 1480c1**
Fixes #1369 

Line 360 `if ! check_hash "$archive" "$hash"; then` was always skipping curl checks (If modified etc) so old patches were never updated

Use mktemp to create a temp file to reside in memory and compare with diff

Changed do_patch's default return code to 1 and only succeed if patch or git am actually worked

### CODE REVIEW

File: build/media-suite_helper.sh [hunk 1]
```diff
@@ -1080,21 +1090,19 @@ do_removeOptions() {
...
-        if [[ -z $patchName ]]; then
-            echo -e "${red}Failed to apply patch '$patch'${reset}"
-            echo -e "${red}Patch without filename, ignoring. Specify an explicit filename.${reset}"
-            return 1
-        fi
-    fi
+    # hack for URLs without filename
+    patchName=${patchName:-"$(/usr/bin/curl -sI "$patch" | grep -Eo 'filename=.*$' | sed 's/filename=//')"}
+    [[ -z $patchName ]] &&
+        printf '%b\n' "${red}Failed to apply patch '$patch'${reset}" \
```
Comment (wiiaboo) [inline build/media-suite_helper.sh]: If the second line's also red, there's no need to reset, right?

Comment (1480c1) [reply build/media-suite_helper.sh]: Remove extra reset and red

**Comment by wiiaboo** [comment]
This is breaking xvidcore (and maybe others) because $nocd is now always defined. You need to refactor do_extract's $nocd check for true/false too or use a parameter like the others to avoid using the parents' $nocd.

**Comment by 1480c1** [comment]
I see, will make sure again, my builds don't seem to build xvidcore, need to make sure

**Comment by wiiaboo** [comment]
Reproducible with `do_wget -h 165ba6a2a447a8375f7b06db5a3c91810181f2898166e7c8137401d7fc894cf0 \
        "https://downloads.xvid.com/downloads/xvidcore-1.3.5.tar.gz" "xvidcore.tar.gz"` (should end up in /build/xvidcore, but it ends up in /build)

**Comment by wiiaboo** [comment]
Yeah, I'm testing this together with the ccache branch, since this one's required for svt-av1 to build.

**Comment by 1480c1** [comment]
Should be fixed

**Comment by 1480c1** [comment]
Wait, nvm

**Comment by 1480c1** [comment]
! wait, nvm

I jebaited myself

**Comment by wiiaboo** [comment]
The `git apply --check` test is not working for me, but the `git am` command works.

**Comment by wiiaboo** [comment]
Also, you removed the check for http/ftp, so curl is trying to download local patches uselessly.

**Comment by 1480c1** [comment]
> The `git apply --check` test is not working for me, but the `git am` command works.

Can you provide a reproducible one? `git apply --check` should be the same functionality as `patch --dry-run` but with git diffs



> Also, you removed the check for http/ftp, so curl is trying to download local patches uselessly.

Need to consider that. Might add the logic to do_wget to test if "url" is a local file

**Comment by wiiaboo** [comment]
It's a local-only patch I created to fix dailymotion URLs. I'm not sure if
it's a valid fix so that's why I didn't submit it to FFmpeg.
When I get home again I'll upload it somewhere.

**Comment by wiiaboo** [comment]
> > The `git apply --check` test is not working for me, but the `git am` command works.
> 
> Can you provide a reproducible one? `git apply --check` should be the same functionality as `patch --dry-run` but with git diffs

https://i.fsbn.eu/ffmpeg-fix-url-hash.patch

**Comment by 1480c1** [comment]
Try again, I added `--ignore-space-change --ignore-whitespace` to match git am's option, although I think having both is redundant since they have the share the same help text in `--help` but are listed separately in the heading portion

**Comment by wiiaboo** [comment]
Merged in 7b5428ba3822f7c7ae65be9ee30eeb62e77b1907

**Pull request closed without merging**